src/pages/Environments.js

```diff
@@ -135,24 +136,33 @@ const Environments = () => {
 
   const refreshData = Utils.withBusyState(setLoading, async () => {
     const creator = getUser().email
+
+    const startTimeForLeoCallsEpochMs = new Date().getTime()
     const [newRuntimes, newDisks, newApps] = await Promise.all([
       Ajax(signal).Runtimes.listV2(shouldFilterRuntimesByCreator ? { creator, includeLabels: 'saturnWorkspaceNamespace,saturnWorkspaceName' } : { includeLabels: 'saturnWorkspaceNamespace,saturnWorkspaceName' }),
       Ajax(signal).Disks.list({ creator, includeLabels: 'saturnApplication,saturnWorkspaceNamespace,saturnWorkspaceName' }),
       Ajax(signal).Apps.listWithoutProject({ creator, includeLabels: 'saturnWorkspaceNamespace,saturnWorkspaceName' })
     ])
+    const endTimeForLeoCallsEpochMs = new Date().getTime()
 
+    const startTimeForRawlsCallsEpochMs = new Date().getTime()
     const decorateLabeledCloudObjWithWorkspaceId = withErrorIgnoring(async cloudObject => {
       const { labels: { saturnWorkspaceNamespace, saturnWorkspaceName } } = cloudObject
       const details = !!saturnWorkspaceNamespace && !!saturnWorkspaceName ? await Ajax(signal).Workspaces.workspace(saturnWorkspaceNamespace, saturnWorkspaceName).details(['workspace']).catch(_ => undefined) : undefined
       const workspaceId = details?.workspace?.workspaceId
       return { ...cloudObject, workspaceId }
     })
 
-    const [decoratedRuntimes, decoratedDisks, decoratedApps] = [
-      await Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newRuntimes)),
-      await Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newDisks)),
-      await Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newApps))
-    ]
+    const [decoratedRuntimes, decoratedDisks, decoratedApps] = await Promise.all([
+      Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newRuntimes)),
+      Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newDisks)),
+      Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newApps))
+    ])
+    const endTimeForRawlsCallsEpochMs = new Date().getTime()
+
+    const leoCallTimeTotalMs = endTimeForLeoCallsEpochMs - startTimeForLeoCallsEpochMs
+    const rawlsCallTimeTotalMs = endTimeForRawlsCallsEpochMs - startTimeForRawlsCallsEpochMs
+    Ajax().Metrics.captureEvent(Events.cloudEnvironmentDetailsLoad, { leoCallTimeMs: leoCallTimeTotalMs, rawlsCallTimeMs: rawlsCallTimeTotalMs, totalCallTimeMs: leoCallTimeTotalMs + rawlsCallTimeTotalMs })
```

Comment:
I wonder if it may be worth having a helper for stopwatching, perhaps a wrapper like

```javascript
// Helper wrapper
const withStopwatch = fn => {
  const startTime =  Date.now
  const res = fn()
  const endTime = Date.now
  const timeElapsed = endTime - startTime

  return {res, elapsedTime}
}

// Main
const {res: [decoratedRuntimes, decoratedDisks, decoratedApps], leoCallTimeMs } = withStopwatch( () => 
  await Promise.all([
      Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newRuntimes)),
      Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newDisks)),
      Promise.all(_.map(decorateLabeledCloudObjWithWorkspaceId, newApps))
    ])
)

Ajax().Metrics.captureEvent(Events.cloudEnvironmentDetailsLoad, { leoCallTimeMs })
```

Reply:
Not opposed to it if it becomes more widespread, but seems like an over-eager generalization that adds complication

Reply:
My main motivation was due to start/stop's sprinkled around looking harder to me to discern the business logic, rather than factoring out duplicate code but I don't feel strongly.

Reply:
I like Kaan's idea, it might look a bit cleaner.